### PR TITLE
RFC: ipq807x: add FAN and LED Support for ZyXEL NBG7815

### PR DESCRIPTION
**Pull request by ddimension**
The device has a TI LP5569 LED controller. This adds support for this driver as a module.

Also there's a FAN. Based on the work on mvebu, I have implemented a simple control for it.

Please give me you opinion about this.

### CODE REVIEW

File: package/kernel/linux/modules/leds.mk [hunk 1]
```diff
@@ -131,6 +131,22 @@ endef
 $(eval $(call KernelPackage,leds-apu))
 
 
+define KernelPackage/leds-lp55xx-common
+  SUBMENU:=$(LEDS_MENU)
+  TITLE:=LED driver common code for LP55XX controllers
+  #DEPENDS:=+kmod-i2c-core +kmod-regmap-i2c
```
Comment (robimarko) [inline package/kernel/linux/modules/leds.mk:137]: There are dependencies in the upstream KConfig, so they should be listed

File: package/kernel/linux/modules/leds.mk [hunk 2]
```diff
@@ -131,6 +131,22 @@ endef
...
+  KCONFIG:=CONFIG_LEDS_LP55XX_COMMON
+  FILES:=$(LINUX_DIR)/drivers/leds/leds-lp55xx-common.ko
+  AUTOLOAD:=$(call AutoLoad,50,leds-lp55xx-common,1)
```
Comment (robimarko) [inline package/kernel/linux/modules/leds.mk:140]: This looks wrong, why is it autoloaded at all and why is it required in preinit?
It should get autoloaded by modules that consume symbols from it anyway.

File: package/kernel/linux/modules/leds.mk [hunk 3]
```diff
@@ -131,6 +131,22 @@ endef
 $(eval $(call KernelPackage,leds-apu))
 
 
+define KernelPackage/leds-lp55xx-common
```
Comment (robimarko) [inline package/kernel/linux/modules/leds.mk:134]: In my opinion, this should be hidden and just selected by individual drivers

**Comment by motolav** [comment]
The driver code taken from the Zyxel sources(in your PR) isn't appropriate for upstream Linux or OpenWrt, I brought it up in the forum thread but I guess it was ignored.

File: package/kernel/leds-lp5569/src/leds-lp5569.c [hunk 4]
```diff
@@ -0,0 +1,1480 @@
...
+#define LP5569_RUN_ENG2			0x20
+#define LP5569_RUN_ENG3			0x08
+
+#define LED_ACTIVE(mux, led)		(!!(mux & (0x0001 << led)))
+
+enum lp5569_chip_id {
+	LP5569,
+};
+
+#if defined(CONFIG_LEDS_LP5569_PREDEFINED_PATTERNS)
```
Comment (motolav) [inline package/kernel/leds-lp5569/src/leds-lp5569.c:133]: All these predefined program patterns must be removed before submitting to the Linux kernel, everything from line 133 to 377.

**Comment by ddimension** [comment]
> Please, no more downstream drivers, this should go upstream first and then it can be backported

As motolav told, this is a device specific driver. I do not know how to work around this, this is definetely to much for me and I would abandon the LED driver.

So at the end, there's only the fan left.  In my personal use here, I already heard it multiple times so it seems to be important. 

What's you opinion about that implementation?
At first I thought about coolingmap and gpio-fan. But I think the user-space will give more flexibility.

File: package/kernel/leds-lp5569/src/leds-lp5569.c (hunk 4 above)
Comment (ddimension) [reply package/kernel/leds-lp5569/src/leds-lp5569.c:133]: Thank you for this good hint. I did not notice Zyxel made changes to it.

**Comment by robimarko** [comment]
But the part isn't device specific at all, it's just off-the-shelf TI LP5596 with a public datasheet meaning its a perfect candidate for upstreaming, otherwise its gonna get left for OpenWrt to maintain pretty much as long as the target is alive.

The issue I see with fan scripts is that we are just gonna have more and more devices with fans, this one is just ON/OFF as far as I can see, so why not just add an active trip point instead of a passive one in DTS and use the gpio-fan device as cooling device instead of CPUFreq, CPUFreq can then be used to reduce clocks if active cooling doesn't help.

**Comment by motolav** [comment]
> > Please, no more downstream drivers, this should go upstream first and then it can be backported
> 
> As motolav told, this is a device specific driver. I do not know how to work around this, this is definetely to much for me and I would abandon the LED driver.

It's not device specific just zyxel made a minor addition but to actually use the driver to mimic the Zyxel firmware would require a userspace script to reprogram the controller to what color/script to run (Based on the kernel docs for the other TI chips)

**Comment by robimarko** [comment]
What has changed in this push?

**Comment by mrkiko** [comment]
Hi!!

Thank for your work again :)

Do you plan to look again at fan support at some point ?

Thanks!

**Pull request closed without merging**